Clojure SDK

JeremS opened the pull request:
Hi this is the clojure SDK. I have been trying to follow what other SDKs are doing.

- there are snippets in the docs
- consts are generated
- I have made the hello world example (and used the generated html)
- I have added the SDK in the list below the others.

I hope I didn't forget anything.

Cheers,

JeremS committed:
Feature: added clojure sdk

Chore: rebasing on latest
(last commit from clojure-sdk d6259824908c)

bencroker commented on build/run.go:
Please put in alphabetical order along with the other SDKs.

bencroker commented on site/static/md/reference/sdks.md:
Order alphabetically, please.

bencroker commented:
Review please, @andersmurphy.

JeremS committed:
Fix: alphabetical order

andersmurphy commented on sdk/clojure/README.md:
```diff
@@ -0,0 +1,178 @@
+
+## Installation
+
+For now the libraries are distributed as git dependencies. You need to add a dependency
+for each library you use.
```
Nice that dependencies are modular and opt in.

andersmurphy commented on examples/clojure/hello-world/src/main/example/server.clj:
```diff
@@ -0,0 +1,32 @@
+    [example.core :as c]
+    [ring.adapter.jetty :as jetty])
+  (:import
+    org.eclipse.jetty.server.Server))
+
+
+
+
+(defonce !jetty-server (atom nil))
+
+
```
This big expanses of whitespace account for a lot of the line count. Some of the files are 50% white space. 😄  

But, do whatever makes it easier for you to maintain/read.

andersmurphy commented on sdk/clojure/README.md:
```diff
@@ -0,0 +1,178 @@
+       (swap! !connections disj sse-gen))}))
+
+
+(defn broadcast-fragment! [fragment]
+  (doseq [c @!connections]
+    (d*/merge-fragment! c fragment)))
```
Nice you got a broadcast example in here.

andersmurphy commented on sdk/clojure/README.md:
```diff
@@ -0,0 +1,178 @@
+- Streamlined release process (cutting releases and publish jar to a maven repo)
+- consider using a byteArrayBuilder (may require adding a dependency)
+- consider uniformizing the adapters behavior on connection closing (throwing in all adapters?)
+- add license files ?
```
Probably one for @bencroker and @delaneyj not sure what the sdk licensing strategy is.

bencroker replied:
It’ll have to be MIT if it is merged to this repo: https://github.com/starfederation/datastar?tab=MIT-1-ov-file#readme

JeremS replied:
No problem with MIT. I believe it should be fine not to add copies of the licence file as long as the SDK is distributed from the main repo as a git dependency. The main repo already provides the license.

When I'll look into distributing jars, I'll include a license file there. I left the todo in the readme for that reason.

bencroker replied:
Thanks!

andersmurphy commented on sdk/clojure/adapter-http-kit/src/main/starfederation/datastar/clojure/adapter/http_kit/impl.clj:
```diff
@@ -0,0 +1,44 @@
+    (let [event-str (-> (StringBuilder.)
+                        (sse/write-event! event-type data-lines opts)
+                        str)]
+      (u/lock! lock
+        (hk-server/send! ch event-str false))))
```
Does http-kit not handle locking?

JeremS replied:
You are right, I didn't source dived Http-kit's code too much but this locking is probably redondant. It allows for preventing the interleaved sending of events from multiple threads like this:

```clojure
(lock! (:lock sse)
  (d*/merge-fragment sse ...)
  (d*/merge-fragment sse ...)
  (d*/merge-signals sse ...))
```
Which may not be necessary. I can remove the lock if you think it's best.

andersmurphy replied:
I think it's probably fine to keep the lock there as it makes the code more consistent with the other adapters (means less depending on http-kits internals). If it stops interleaved sending that's probably a good thing too.

andersmurphy commented on sdk/clojure/adapter-jetty/src/main/starfederation/datastar/clojure/adapter/ring_jetty/impl.clj:
```diff
@@ -0,0 +1,59 @@
+
+(defrecord SSE-gen [writer lock on-open on-close]
+  rp/StreamableResponseBody
+  (write-body-to-stream [this _ output-stream]
+    (u/lock! lock
+      (when (deref writer)
+        (throw (ex-info "Reused SSE-gen as several ring responses body. Don't do this." {})))
+      (.flush ^OutputStream output-stream)
+      (vreset! writer (io/writer output-stream)))
+    (on-open this))
```
Didn't realize the output stream didn't lock on write. So is this to prevent partial interleaved messages from multiple writers (cause people will do crazy things with the sdk)?

JeremS replied:
The design document for the SDK talks of locking, so I put it place just to be sure. I went a bit grug at the start. 

Here the lock serves several purposes:
- the code you cite prevents from different threads using the same sse generator for different responses bodies. Since it is possible to do and I don't think this is a good Idea I did it that way. 
- the lock is also used when writing to the bufferedWriter. It would make sense that writing the whole event directly to the output stream is thread safe. But since I use a buffered writer and that I don't write to it in one go, I felt I needed to protect from concurrent writes.
- the lock is last used when closing the buffered writer.
- same as for Http-kit it allows for holding the lock while sending several events.

I don't know if it is the best approach, I may be able to use a new bufferedWriter each time the `send-event!` function is called instead of holding on the same one. What do you think? This is hard the part of SDK and I really wanna get it right.

andersmurphy replied:
Honestly, I'm not sure. I do use one buffered writer for each connection (as I've hooked in compression). But, my set up also doesn't allow for concurrent writes for a connections (as it's all driven by a render function that's basically a queue), but as an sdk you can't have those guarantees.

 I do think it's nice to be able to hold the lock for sending multiple events for that user as it gives you atomic batches which is nice. So I say keep it the way you have it.

sundbp replied:
Agree. Less surprises in concurrent situations is a good thing. And the overhead of the lock here is likely unimportant, or not important enough to care about before it is proven important enough.

andersmurphy commented on sdk/clojure/deps.edn:
```diff
@@ -0,0 +1,38 @@
+                       :test.paths/core-sdk
+                       :test.paths/malli-schemas
+                       :test.paths/adapter-common
+                       :test.paths/adapter-http-kit
+                       :test.paths/adapter-ring-jetty]
+
+
+         :extra-deps {io.github.noahtheduke/lazytest {:mvn/version "1.5.0"}
+                      metosin/reitit                 {:mvn/version "0.7.2"}
+                      etaoin/etaoin                  {:mvn/version "1.1.42"}
```
Wow etoain based UI smoketests?

JeremS replied:
Yeah the tests with it are basic for now. Still it is nice to be able to automate some end to end examples and it's already there if its needed more extensively later. Also it's fun when it works!

Superpat commented:
@JeremS Did you test out the shell based sdk test suite found in sdk/test ?

Superpat commented:
It's currently a bit too strict about dataline ordering, but besides that should work well.

andersmurphy commented on sdk/clojure/doc/implementing-adapters.md:
````diff
@@ -0,0 +1,110 @@
+
+```
+
+As per the design doc that all Datastar SDKs follow, we use a lock in this
+function (`java.util.concurrent.locks.ReentrantLock`) to protect
+from several threads concurrently writing any underlying buffer before flushing.
````
Today I learned.

JeremS replied:
See the comment about the jetty adapter, I think you can interleave writings to a bufferedWriter and since I hold onto one in the generator...

andersmurphy commented on sdk/clojure/sdk/src/main/starfederation/datastar/clojure/api/fragments.clj:
```diff
@@ -0,0 +1,198 @@
+
+
+(defn remove-fragment! [sse-gen selector opts]
+  (when-not (valid-selector? selector)
+    (throw (ex-info "Invalid selector" {:selector selector})))
```
Liking the structured errors.

JeremS replied:
I copied at lot of the go sdk for this. I figured all the relevant error to catch and distinguish between have already been thought out there. I ported it to clojure...

andersmurphy commented on sdk/clojure/sdk/src/main/starfederation/datastar/clojure/api.clj:
```diff
@@ -0,0 +1,411 @@
+  (p/close-sse! sse-gen))
+
+
+(defmacro with-open-sse
```
Great to have a with macro!

andersmurphy reviewed (approved):
This is a brilliant. Love the amount of tests, examples, docs, etc. You've really done an extensive piece of work here. Just a few comments but nothing major. 🚀

JeremS commented:
@Superpat Sorry I didn't yet, but I'll definitely tackle that next. Since I have some test of my own I wanted to put the code out there first. I felt I needed other sets of eyeballs than my own at this point, it was time to pull the trigger and at least create this PR.

JeremS committed:
Docs: Actual broadcast example

bencroker commented:
This is looking great, @JeremS! I’m marking as draft while you work through things with @andersmurphy and @Superpat. Please mark as ready for review when you feel it is ready for merging.

JeremS commented:
@bencroker Thank you! I'll figure out a way to hook up into @Superpat tests then I'll mark ready for review!

Superpat commented:
@JeremS no problem. If you notice that a case your testing for in your existing tests but that's missing in the shell based tests let me know.

sundbp commented on sdk/clojure/sdk/src/main/starfederation/datastar/clojure/api/sse.clj:
```diff
@@ -0,0 +1,140 @@
+    java.lang.Appendable))
+
+
+(def SSE-headers-1
```
1 based index instead of 0 based? :)

JeremS replied:
not really, it more about HTTP version number...

bencroker commented:
Please merge `develop` and use `helloWorldExample` directly:
https://github.com/starfederation/datastar/blob/a2e4181450b92fc04cf0a654204a1a7846c2baab/build/run.go#L150

JeremS committed:
Test: adding shell test support

Fix: order of options in data-lines for execute scripts events

Fix: fixed signal handling / missing options

Refactor: using the generated constants for boolean options

JeremS committed:
Merge branch 'develop' into clojure-sdk-release

JeremS committed:
Test: passing shell tests

Chore: task build

bencroker commented:
Looking good, can you please resolve or respond to all open conversations?

A final review by @andersmurphy would be great, too!

andersmurphy reviewed (approved):
LGTM! Nice one! 🚀

bencroker commented:
Merged, thanks!
